addons/mail/tests/discuss/test_discuss_mail_presence.py

```diff
@@ -15,86 +17,59 @@
-        channel._add_members(guests=other_guest)
-        # Now that they share a channel, guest should receive guest's presence.
-        self._receive_presence(sender=other_guest, recipient=guest)
-        self.assertEqual(other_guest.im_status, "online")
-
-    def test_receive_presences_as_portal(self):
-        portal = new_test_user(self.env, login="portal_user", groups="base.group_portal")
-        bob = new_test_user(self.env, login="bob_user", groups="base.group_user")
-        # Portal should not receive users's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=bob, recipient=portal)
-        channel = self.env["discuss.channel"]._create_channel(group_id=None, name="General")
-        channel._add_members(users=portal | bob)
-        # Now that they share a channel, portal should receive users's presence.
-        self._receive_presence(sender=bob, recipient=portal)
-
-        guest = self.env["mail.guest"].create({"name": "Guest"})
-        # Portal should not receive guest's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=guest, recipient=portal)
-        channel._add_members(guests=guest)
-        # Now that they share a channel, portal should receive guest's presence.
-        self._receive_presence(sender=guest, recipient=portal)
-        self.assertEqual(guest.im_status, "online")
-
-    def test_receive_presences_as_internal(self):
+    def test_presence_access(self):
         internal = new_test_user(self.env, login="internal_user", groups="base.group_user")
+        other_internal = new_test_user(
+            self.env, login="other_internal_user", groups="base.group_user"
+        )
+        portal = new_test_user(self.env, login="portal_user", groups="base.group_portal")
+        other_portal = new_test_user(
+            self.env, login="other_portal_user", groups="base.group_portal"
+        )
         guest = self.env["mail.guest"].create({"name": "Guest"})
-        # Internal can access guest's presence regardless of their channels.
-        self._receive_presence(sender=guest, recipient=internal)
-        # Internal can access users's presence regardless of their channels.
-        bob = new_test_user(self.env, login="bob_user", groups="base.group_user")
-        self._receive_presence(sender=bob, recipient=internal)
-        self.assertEqual(bob.im_status, "online")
+        other_guest = self.env["mail.guest"].create({"name": "Other Guest"})
+        for requested_by, target, has_token, allowed in [
+            *product([internal], [guest, other_internal, portal], [True, False], [True]),
+            *product([guest, portal], [internal, other_guest, other_portal], [False], [False]),
+            *product([guest, portal], [internal, other_guest, other_portal], [True], [True]),
```

Comment:
I guess we gain lines but in terms of clarity I think the main issue is the lack of named params.
The expanded list had the same issue anyway.

Considering all the tests we have with such a pattern, we might want to create a custom class that does the job. But for another PR.